gui/test/app/tinystan/model.test.ts

```diff
@@ -0,0 +1,427 @@
+          metric[0][0][0] + numFreeParams * numFreeParams,
+        );
+      }
+    });
+
+    test("seeding works", async () => {
+      const { mockedModule, model } = await getMockedModel({});
+
+      const seed = 123;
+      model.sample({ seed });
+
+      expect(mockedModule._tinystan_sample).toHaveBeenNthCalledWith(
+        1,
+        ...Array(3).fill(expect.any(Number)),
+        seed,
+        ...Array(24).fill(expect.any(Number)),
+      );
+    });
+
+    test("errors on bad output sizes", async () => {
+      const { mockedModule, model } = await getMockedModel({});
+      expect(() => model.sample({ num_chains: 0 })).toThrow(/num_chains/);
+      expect(() => model.sample({ num_samples: 0 })).toThrow(/num_samples/);
+      expect(() => model.sample({ num_warmup: -1 })).toThrow(/num_warmup/);
+
+      expect(mockedModule._tinystan_sample).toHaveBeenCalledTimes(0);
+    });
+  });
+
+  describe("pathfinder function", () => {
+    test("null call behavior", async () => {
+      const { mockedModule, model } = await getMockedModel({});
+      model.pathfinder({});
+
+      expect(mockedModule._tinystan_pathfinder).toHaveBeenCalledTimes(1);
+    });
+
+    test("erroring call behavior", async () => {
+      const { mockedModule, model } = await getMockedModel({ returnCode: 1 });
+      expect(() => model.pathfinder({})).toThrow(/Error inside WASM/);
+
+      expect(mockedModule._tinystan_pathfinder).toHaveBeenCalledTimes(1);
+    });
+
+    test("empty model errors", async () => {
+      const { mockedModule, model } = await getMockedModel({
+        numFreeParams: 0,
+      });
+      expect(() => model.pathfinder({})).toThrow(/no parameters/);
+
+      expect(mockedModule._tinystan_pathfinder).toHaveBeenCalledTimes(0);
+    });
+
+    test("model failing to allocate errors", async () => {
+      const { mockedModule, model, getStdout } = await getMockedModel({
+        modelFails: true,
+      });
+      expect(() => model.pathfinder({})).toThrow(/Error inside WASM/);
+
+      expect(getStdout()).toMatch(/Error inside WASM/);
+      expect(mockedModule._tinystan_pathfinder).toHaveBeenCalledTimes(0);
+    });
+
+    test("data can be string or record", async () => {
+      const { mockedModule, model } = await getMockedModel({});
+
+      const data = { hi: 1 };
+      const dataStr = JSON.stringify(data);
+
+      model.pathfinder({ data });
+      expect(mockedModule.stringToUTF8).toHaveBeenNthCalledWith(
+        1,
+        dataStr,
+        expect.any(Number),
+        expect.any(Number),
+      );
+
+      mockedModule.stringToUTF8.mockClear();
+
+      model.pathfinder({ data: dataStr });
+      expect(mockedModule.stringToUTF8).toHaveBeenNthCalledWith(
+        1,
+        dataStr,
+        expect.any(Number),
+        expect.any(Number),
+      );
+
+      expect(mockedModule._tinystan_pathfinder).toHaveBeenCalledTimes(2);
+    });
+
+    test("output is correct shape and layout", async () => {
```

Comment:
Also looks like this could be a loop, if you don't mind some conditionals in the loop.